### Import Libaries:

In [75]:
import numpy as np
import pandas as pd
import scipy as sp
from scipy.stats import mode
from sklearn.linear_model import LinearRegression as LinReg
import matplotlib
import matplotlib.pyplot as plt
from sklearn import discriminant_analysis
from sklearn.decomposition import PCA
from sklearn import preprocessing
from collections import Counter
%matplotlib inline

### Import listings, clean data, extract features

In [2]:
# Read in the data 
listings = pd.read_csv('listings.csv', delimiter=',')

# View feature list
print listings.columns.values

Fig 1

['id' 'scrape_id' 'last_scraped' 'name' 'picture_url' 'host_id' 'host_name'
 'host_since' 'host_picture_url' 'street' 'neighbourhood'
 'neighbourhood_cleansed' 'city' 'state' 'zipcode' 'market' 'country'
 'latitude' 'longitude' 'is_location_exact' 'property_type' 'room_type'
 'accommodates' 'bathrooms' 'bedrooms' 'beds' 'bed_type' 'square_feet'
 'price' 'weekly_price' 'monthly_price' 'guests_included' 'extra_people'
 'minimum_nights' 'maximum_nights' 'calendar_updated' 'availability_30'
 'availability_60' 'availability_90' 'availability_365'
 'calendar_last_scraped' 'number_of_reviews' 'first_review' 'last_review'
 'review_scores_rating' 'review_scores_accuracy'
 'review_scores_cleanliness' 'review_scores_checkin'
 'review_scores_communication' 'review_scores_location'
 'review_scores_value' 'host_listing_count']


In [46]:
print 'Listings Data Shape: ', listings.shape

# Unsuppress Output
pd.options.display.max_columns = 77

listings.head(n = 3)

Listings Data Shape:  (27392, 52)


id     scrape_id last_scraped                                 name  \
0  1069266  2.015010e+13       1/2/15        Stay like a real New Yorker!    
1  1846722  2.015010e+13       1/2/15    Apartment 20 Minutes Times Square   
2  2061725  2.015010e+13       1/2/15  Option of 2 Beds w Private Bathroom   

                                         picture_url  host_id host_name  \
0  https://a0.muscache.com/pictures/50276484/larg...  5867023   Michael   
1  https://a1.muscache.com/pictures/35865039/larg...  2631556    Denise   
2  https://a2.muscache.com/pictures/50650147/larg...  4601412      Miao   

  host_since                                   host_picture_url  \
0    4/10/13  https://a2.muscache.com/ic/users/5867023/profi...   
1    6/13/12  https://a2.muscache.com/ic/users/2631556/profi...   
2     1/5/13  https://a0.muscache.com/ic/users/4601412/profi...   

                                              street     neighbourhood  \
0  East 53rd Street, New York, NY 10022, United S...      Midtown East   
1     West 155th Street, New York, NY, United States  Hamilton Heights   
2  Van Buren Street, Brooklyn, NY 11221, United S...          Bushwick   

  neighbourhood_cleansed      city state     zipcode    market        country  \
0           Midtown East  New York    NY  10022-4175  New York  United States   
1       Hamilton Heights  New York    NY         NaN  New York  United States   
2               Bushwick  Brooklyn    NY       11221  New York  United States   

    latitude  longitude is_location_exact property_type        room_type  \
0  40.756852 -73.964754                 t     Apartment  Entire home/apt   
1  40.830599 -73.941014                 f     Apartment  Entire home/apt   
2  40.692189 -73.924120                 t     Apartment     Private room   

   accommodates  bathrooms  bedrooms  beds  bed_type  square_feet    price  \
0             2        1.0       1.0   1.0  Real Bed          NaN  $160.00   
1            10        1.0       3.0   3.0  Real Bed          NaN  $105.00   
2             2        1.0       1.0   2.0  Real Bed          NaN   $58.00   

  weekly_price monthly_price  guests_included extra_people  minimum_nights  \
0    $1,000.00           NaN                2        $0.00               3   
1          NaN           NaN                1       $50.00               1   
2          NaN           NaN                1       $12.00               3   

   maximum_nights calendar_updated  availability_30  availability_60  \
0              14      3 weeks ago               21               51   
1             180       4 days ago               28               58   
2              30            today                4               13   

   availability_90  availability_365 calendar_last_scraped  number_of_reviews  \
0               72               322                1/2/15                 62   
1               88               348                1/2/15                 22   
2               26               227                1/2/15                 35   

  first_review last_review  review_scores_rating  review_scores_accuracy  \
0      4/28/13    12/17/14                  86.0                     9.0   
1       1/5/14    12/29/14                  85.0                     8.0   
2       2/4/14    12/29/14                  98.0                    10.0   

   review_scores_cleanliness  review_scores_checkin  \
0                        7.0                    9.0   
1                        8.0                    9.0   
2                       10.0                   10.0   

   review_scores_communication  review_scores_location  review_scores_value  \
0                          9.0                    10.0                  9.0   
1                          8.0                     7.0                  8.0   
2                         10.0                     9.0                 10.0   

   host_listing_count  
0                   1  
1                   2  
2                   4

For our baseline model, we can start by using features that we intuitively sense will impact a listing's price. Here are some good features to start with.
- price
- neighborhood_cleansed
- property_type
- room_type
- accommodates
- bathrooms
- bedrooms
- guests_included
- extra_people
- number_of_reviews
- review_scores_rating
- host_listing_count

This brings our feature list down from 52 to 13.



In [64]:
listings_slim = listings[['property_type',
                  'price',
                  'neighbourhood_cleansed',
                  'room_type',
                  'accommodates',
                  'bathrooms',
                  'bedrooms',
                  'beds',
                  'guests_included',
                  'extra_people',
                  'number_of_reviews',
                  'review_scores_rating',
                  'host_listing_count']]

print 'Size of trimmed data: ', listings_slim.shape

Size of trimmed data:  (27392, 13)


Let's remove entries (rows) that have faulty data like when
- There are 0 bedrooms
- There are 0 bathrooms
- There are 0 beds
- The price is $0

In [65]:
# Delete bad entries
listings_slim = listings_slim[listings_slim.bedrooms != 0]
listings_slim = listings_slim[listings_slim.beds != 0]
listings_slim = listings_slim[listings_slim.price != 0]

# Delete additional entries with NaN values
listings_slim = listings_slim.dropna(axis=0)

print 'Size of trimmed data: ', listings_slim.shape

Size of trimmed data:  (16694, 13)


We also need to drop the dollar sign from our price and turn the type into a float.

In [66]:
# Convert $ to float for 'price'
listings_slim['price'] = listings_slim['price'].replace('[\$,)]','',  \
        regex=True).replace('[(]','-', regex=True).astype(float)

# Convert $ to float for 'extra people'
listings_slim['extra_people'] = listings_slim['extra_people'].replace('[\$,)]','',  \
        regex=True).replace('[(]','-', regex=True).astype(float)

Because we are doing OLS for our baseline regression, we must have only numerical predictors and so we must also one-hot encode our categorical variables.

In [67]:
def one_hot(x_df):
    x_df_expanded = pd.DataFrame({})
    
    # From lecture notes to remove categorical
    for column in x_df.columns:
        # Check if attribute is not categorical: either dtype is not object
        if(x_df[column].dtype == np.dtype('object') or x_df[column].dtype == np.dtype('string')):
            # otherwise: use one-hot encoding
            encoding = pd.get_dummies(x_df[column], prefix=column)  # Convert categorical variable into dummy/indicator variables
            # append expanded attribute to data frame
            x_df_expanded = pd.concat([x_df_expanded, encoding], axis=1)
        else:
            # standardizes non-categorial variables
#             x_df[column] = x_df[column] / x_df[column].std(axis = 0)

            x_df_expanded = pd.concat([x_df_expanded, x_df[column]], axis=1)
    return x_df_expanded

print 'Size of trimmed data: ', listings_slim.shape
listings_slim = one_hot(listings_slim)
print 'Size of trimmed data: ', listings_slim.shape

listings_slim.head(n=4)

Size of trimmed data:  (16694, 13)
Size of trimmed data:  (16694, 197)


property_type_Apartment  property_type_Bed & Breakfast  property_type_Boat  \
0                      1.0                            0.0                 0.0   
1                      1.0                            0.0                 0.0   
2                      1.0                            0.0                 0.0   
3                      1.0                            0.0                 0.0   

   property_type_Cabin  property_type_Castle  property_type_Chalet  \
0                  0.0                   0.0                   0.0   
1                  0.0                   0.0                   0.0   
2                  0.0                   0.0                   0.0   
3                  0.0                   0.0                   0.0   

   property_type_Dorm  property_type_Earth House  property_type_House  \
0                 0.0                        0.0                  0.0   
1                 0.0                        0.0                  0.0   
2                 0.0                        0.0                  0.0   
3                 0.0                        0.0                  0.0   

   property_type_Lighthouse  property_type_Loft  property_type_Other  \
0                       0.0                 0.0                  0.0   
1                       0.0                 0.0                  0.0   
2                       0.0                 0.0                  0.0   
3                       0.0                 0.0                  0.0   

   property_type_Tent  property_type_Treehouse  property_type_Villa  price  \
0                 0.0                      0.0                  0.0  160.0   
1                 0.0                      0.0                  0.0  105.0   
2                 0.0                      0.0                  0.0   58.0   
3                 0.0                      0.0                  0.0  185.0   

   neighbourhood_cleansed_Allerton  neighbourhood_cleansed_Alphabet City  \
0                              0.0                                   0.0   
1                              0.0                                   0.0   
2                              0.0                                   0.0   
3                              0.0                                   0.0   

   neighbourhood_cleansed_Astoria  neighbourhood_cleansed_Bath Beach  \
0                             0.0                                0.0   
1                             0.0                                0.0   
2                             0.0                                0.0   
3                             0.0                                0.0   

   neighbourhood_cleansed_Battery Park City  neighbourhood_cleansed_Bay Ridge  \
0                                       0.0                               0.0   
1                                       0.0                               0.0   
2                                       0.0                               0.0   
3                                       0.0                               0.0   

   neighbourhood_cleansed_Baychester  neighbourhood_cleansed_Bayside  \
0                                0.0                             0.0   
1                                0.0                             0.0   
2                                0.0                             0.0   
3                                0.0                             0.0   

   neighbourhood_cleansed_Bedford Park  \
0                                  0.0   
1                                  0.0   
2                                  0.0   
3                                  0.0   

   neighbourhood_cleansed_Bedford-Stuyvesant  \
0                                        0.0   
1                                        0.0   
2                                        0.0   
3                                        0.0   

   neighbourhood_cleansed_Bensonhurst  neighbourhood_cleansed_Bergen Beach  \
0                                 0.0                                  0.0   
1          

We can see that one-hot encoding brought our number of features from 13 to 197. In actuality and as a quick sanity check, the only categorical variables are 'property_type', 'room_type', and 'neighbourhood_cleansed' and there are 186 neighbourhoods so this makes sense.

### Baseline OLS Model

In [102]:
# stores our predictors as the independent, x, variables
X = listings_slim.drop('price', axis = 1)

# stores price as the dependent, y, variable
Y = listings_slim['price']

linreg = LinReg()
linreg.fit(X,Y)
training_set_score = linreg.score(X,Y)
print 'The R^2 score on our training data is: ' + str(round(training_set_score,3))

The R^2 score on our training data is: 0.489


In [111]:
# stores the coefficient values of the predictors
coefficient_values = np.array(linreg.coef_)

# stores the names of the variables
variable_names = X.columns.values

# creates table storing the coefficient values and variable names
coef_matrix = pd.DataFrame({'CoefValues':coefficient_values, 'VarName': variable_names, 'AbsCoef': abs(coefficient_values)})

Below is a table that contains the sorted coefficient values for each variable that we decided to include.

In [115]:
sorted_coef_matrix = coef_matrix.sort(columns='AbsCoef').drop('AbsCoef', axis=1)
sorted_coef_matrix

/Users/ChaseDavis/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  if __name__ == '__main__':


CoefValues                                            VarName
192  6.015169e-02                                       extra_people
193 -3.239639e-01                                  number_of_reviews
194  9.691101e-01                               review_scores_rating
195 -1.146117e+00                                 host_listing_count
191  4.755259e+00                                    guests_included
190 -1.045568e+01                                               beds
187  1.388578e+01                                       accommodates
189  5.746099e+01                                           bedrooms
188  6.774525e+01                                          bathrooms
2   -1.957139e+07                                 property_type_Boat
10  -1.957141e+07                                 property_type_Loft
5   -1.957141e+07                               property_type_Chalet
1   -1.957142e+07                      property_type_Bed & Breakfast
3   -1.957142e+07                                property_type_Cabin
8   -1.957143e+07                                property_type_House
12  -1.957143e+07                                 property_type_Tent
13  -1.957143e+07                            property_type_Treehouse
0   -1.957144e+07                            property_type_Apartment
14  -1.957145e+07                                property_type_Villa
11  -1.957146e+07                                property_type_Other
9   -1.957148e+07                           property_type_Lighthouse
4   -1.957149e+07                               property_type_Castle
6   -1.957150e+07                                 property_type_Dorm
7   -1.957151e+07                          property_type_Earth House
162  1.919202e+11                 neighbourhood_cleansed_Tottenville
81   1.919202e+11                    neighbourhood_cleansed_Grasmere
90   1.919202e+11                   neighbourhood_cleansed_Hillcrest
177  1.919202e+11                  neighbourhood_cleansed_Whitestone
30   1.919202e+11                   neighbourhood_cleansed_Bronxdale
26   1.919202e+11                neighbourhood_cleansed_Bergen Beach
53   1.919202e+11                neighbourhood_cleansed_Country Club
103  1.919202e+11                    neighbourhood_cleansed_Longwood
43   1.919202e+11                     neighbourhood_cleansed_Clifton
151  1.919202e+11                  neighbourhood_cleansed_St. George
97   1.919202e+11            neighbourhood_cleansed_Kew Garden Hills
152  1.919202e+11                   neighbourhood_cleansed_Stapleton
63   1.919202e+11               neighbourhood_cleansed_East New York
35   1.919202e+11                    neighbourhood_cleansed_Canarsie
128  1.919202e+11                     neighbourhood_cleansed_Oakwood
112  1.919202e+11              neighbourhood_cleansed_Middle Village
119  1.919202e+11                  neighbourhood_cleansed_Mott Haven
122  1.919202e+11                neighbourhood_cleansed_New Brighton
117  1.919202e+11                 neighbourhood_cleansed_Morris Park
89   1.919202e+11                  neighbourhood_cleansed_Highbridge
146  1.919202e+11                   neighbourhood_cleansed_Soundview
23   1.919202e+11                neighbourhood_cleansed_Bedford Park
96   1.919202e+11                  neighbourhood_cleansed_Kensington
161  1.919202e+11               neighbourhood_cleansed_Tompkinsville
108  1.919202e+11                     neighbourhood_cleansed_Maspeth
50   1.919202e+11           neighbourhood_cleansed_Concourse Village
21   1.919202e+11                  neighbourhood_cleansed_Baychester
110  1.919202e+11              neighbourhood_cleansed_Meiers Corners
60   1.919202e+11               neighbourhood_cleansed_East Elmhurst
157  1.919202e+11               neighbourhood_cleansed_The Rockaways
129  1.919202e+11                  neighbourhood_cleansed_Ozone Park
132  1.919202e+11                  neighbourhood_cleansed_Pelham Bay
131  1.919202e+11             neighbourhood_cleansed_Park Versailles
143  

As we can see, our three categorical features have the same weight despite their encodings – this would likely not be the case in a non-linear model and will be interesting to explore.